xdsl/dialects/experimental/dmp.py

```diff
@@ -0,0 +1,436 @@
+        buff_ub: list[int] = []
+        core_lb: list[int] = []
+        core_ub: list[int] = []
+
+        while True:
+            parser.parse_char("[")
+            buff_lb.append(parser.parse_integer())
+            parser.parse_char(",")
+            core_lb.append(parser.parse_integer())
+            parser.parse_char(",")
+            core_ub.append(parser.parse_integer())
+            parser.parse_char(",")
+            buff_ub.append(parser.parse_integer())
+            parser.parse_char("]")
+            if parser.tokenizer.starts_with("x"):
+                parser.parse_char("x")
+            else:
+                break
+        parser.parse_char(">")
+
+        typ = builtin.i64
+        return [
+            builtin.DenseArrayBase.from_list(typ, data)
+            for data in (buff_lb, buff_ub, core_lb, core_ub)
+        ]
+
+
+@irdl_attr_definition
+class NodeGrid(ParametrizedAttribute):
+    name = "dmp.grid"
+
+    shape: ParameterDef[builtin.DenseArrayBase]
+
+    def __init__(self, shape: Sequence[int]):
+        if len(shape) < 1:
+            raise ValueError("dmp.grid must have at least one dimensions!")
+        super().__init__([builtin.DenseArrayBase.from_list(builtin.i64, shape)])
+
+    def as_tuple(self):
+        shape = self.shape.as_tuple()
+        assert isa(shape, tuple[int])
+        return shape
+
+    @staticmethod
+    def parse_parameters(parser: Parser) -> list[Attribute]:
+        parser.parse_char("<")
+
+        shape: list[int] = [parser.parse_integer(allow_negative=False)]
+
+        while parser.try_parse_characters("x") is not None:
+            shape.append(parser.parse_integer(allow_negative=False))
+
+        parser.parse_char(">")
+
+        return [builtin.DenseArrayBase.from_list(builtin.i64, shape)]
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string("<")
+        printer.print_string("x".join(str(x) for x in self.shape.as_tuple()))
+        printer.print_string(">")
+
+
+@irdl_op_definition
+class HaloSwapOp(IRDLOperation):
+    """
+    Declarative swap of memref regions.
+    """
+
+    name = "dmp.swap"
+
+    input_stencil: Annotated[Operand, stencil.TempType | memref.MemRefType]
+
+    # shape: OptOpAttr[HaloShapeInformation]
+    swaps: OptOpAttr[builtin.ArrayAttr[HaloExchangeDecl]]
+    nodes: OptOpAttr[NodeGrid]
+
+    @staticmethod
+    def get(input_stencil: SSAValue | Operation):
+        return HaloSwapOp.build(operands=[input_stencil])
+
+
+@irdl_op_definition
+class GatherOp(IRDLOperation):
+    """
+    Gather a scattered array back to one node
+    """
+
+    name = "halo.gather"
+
+    local_field: Annotated[Operand, stencil.FieldType | memref.MemRefType]
+    global_field: Annotated[Operand, stencil.FieldType | memref.MemRefType]
+
+    root_rank: OpAttr[builtin.IntegerAttr[builtin.IndexType]]
+
+    retain_order: OptOpAttr[builtin.UnitAttr]
+    """
+    A naive mpi.gather() will result in "pancaking", where each nodes data will
```

Comment:
Avoid using terms like "pancaking".
Maybe add a small pseudocode example

Reply:
Better like this?